src/globus_cli/termio/field.py

```diff
@@ -0,0 +1,110 @@
+
+
+class FieldFormatter(abc.ABC):
+    @abc.abstractmethod
+    def format(self, value: t.Any) -> str | None:
+        ...
+
+
+class _StrFieldFormatter(FieldFormatter):
+    def format(self, value: t.Any) -> str | None:
+        return str(value)
+
+
+class _DateFieldFormatter(FieldFormatter):
+    def format(self, value: t.Any) -> str | None:
+        if not value:
+            return None
+        # let this raise ValueError
+        date = datetime.datetime.fromisoformat(value)
+        if date.tzinfo is None:
+            return date.strftime("%Y-%m-%d %H:%M:%S")
+        return date.astimezone().strftime("%Y-%m-%d %H:%M:%S")
+
+
+class _BoolFieldFormatter(FieldFormatter):
+    def format(self, value: t.Any) -> str | None:
+        return str(bool(value))
+
+
+def _key_to_keyfunc(k):
+    """
+    We allow for 'keys' which are functions that map columns onto value
+    types -- they may do formatting or inspect multiple values on the
+    object. In order to support this, wrap string keys in a simple function
+    that does the natural lookup operation, but return any functions we
+    receive as they are.
+    """
+    # if the key is a string, then the "keyfunc" is just a basic lookup
+    # operation -- return that
+    if isinstance(k, str):
+        subkeys = k.split(".")
+
+        def lookup(x):
+            current = x
+            for sub in subkeys:
+                current = current.get(sub)
+                if current is None:
+                    return None
+            return current
+
+        return lookup
+    # otherwise, the key must be a function which is executed on the item
+    # to produce a value -- return it verbatim
+    return k
+
+
+class Field:
+    """A field which will be shown in record or table output.
+    When fields are provided as tuples, they are converted into this.
+
+    :param name: the displayed name for the record field or the column
+        name for table output
+    :param key: a str for indexing into print data or a callable which
+        produces a string given the print data
+    :param wrap_enabled: in record output, is this field allowed to wrap
+    """
+
+    class FormatName(enum.Enum):
+        Str = enum.auto()
+        Date = enum.auto()
+        Bool = enum.auto()
+
+    _DEFAULT_FORMATTERS: dict[FormatName, FieldFormatter] = {
+        FormatName.Str: _StrFieldFormatter(),
+        FormatName.Date: _DateFieldFormatter(),
+        FormatName.Bool: _BoolFieldFormatter(),
+    }
+
+    def __init__(
+        self,
+        name,
+        key,
+        wrap_enabled=False,
+        formatter: FormatName | FieldFormatter = FormatName.Str,
+    ):
+        self.name = name
+        self.keyfunc = _key_to_keyfunc(key)
+        self.wrap_enabled = wrap_enabled
+
+        if isinstance(formatter, FieldFormatter):
+            self.formatter: FieldFormatter = formatter
+        elif isinstance(formatter, self.FormatName):
+            self.formatter = self._DEFAULT_FORMATTERS[formatter]
```

Comment:
I don't think this is a case where having an Enum adds much value.

It just adds a layer of indirection so someone can refer to a Formatter by name without referring to the formatter itself which unecessarily obfuscates the formatter lookup. Why not just expose a basic list of supported Formatters including primitive ones like Str?

It makes it so that no type unioning/special case lookups need to happen & doesn't hurt any of the user-experience at all imo.

Reply:
I think we got this to a better place after discussion, so I'm marking this resolved. Please reraise this issue if the current contents of the branch don't look right to you though.